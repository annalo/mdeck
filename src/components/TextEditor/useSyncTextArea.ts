import { useEffect } from "react";
import type { RefObject } from "react";
import * as R from "ramda";

interface UseSyncTextAreaProps {
  ref: RefObject<HTMLTextAreaElement>;
  slideshowLineNumber: number;
  textAreaLineHeight: number;
}

export const useSyncTextArea = ({
  ref,
  slideshowLineNumber,
  textAreaLineHeight,
}: UseSyncTextAreaProps): void => {
  /* Syncs text when slideshowLineNumber changes */
  useEffect(() => {
    const getNode = R.prop("current");
    const scrollTop = R.multiply(textAreaLineHeight, slideshowLineNumber);
    // TODO smooth scrolling
    const setScrollTop = R.curry((value, n) => {
      n.scrollTop = value; // eslint-disable-line no-param-reassign
    });

    R.pipe(getNode, R.unless(R.isNil, setScrollTop(scrollTop)))(ref);
  }, [ref, slideshowLineNumber, textAreaLineHeight]);
};
