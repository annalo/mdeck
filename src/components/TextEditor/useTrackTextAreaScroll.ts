import { useEffect, useMemo } from "react";
import * as R from "ramda";
import throttle from "lodash/throttle";

import type { Dispatch, RefObject } from "react";
import {
  MarkdownContextReducerAction,
  MarkdownContextReducerActionType,
} from "types/markdown-context";

interface UseTrackTextAreaScrollProps {
  dispatch: Dispatch<MarkdownContextReducerAction>;
  isActive: boolean;
  ref: RefObject<HTMLTextAreaElement>;
  textAreaLineHeight: number;
}

function useTrackTextAreaScroll({
  dispatch,
  isActive,
  ref,
  textAreaLineHeight,
}: UseTrackTextAreaScrollProps): void {
  const handleScroll = useMemo(
    () =>
      throttle((e) => {
        const getScrollTop = R.path(["target", "scrollTop"]);
        const calculateLineNumber = R.divide(R.__, textAreaLineHeight);
        const textLineNumber = R.pipe(
          getScrollTop,
          calculateLineNumber,
          Math.floor
        )(e);

        dispatch({
          type: MarkdownContextReducerActionType.SetTextLineNumber,
          textLineNumber,
        });
      }, 200),
    [dispatch, textAreaLineHeight]
  );

  /* Adds/Removes event listener on 'scroll' depending on pane `isActive` */
  useEffect(() => {
    const node = ref.current;

    isActive
      ? node?.addEventListener("scroll", handleScroll, { passive: true })
      : node?.removeEventListener("scroll", handleScroll);

    return () => node?.removeEventListener("scroll", handleScroll);
  }, [isActive, handleScroll, ref]);
}

export { useTrackTextAreaScroll };
