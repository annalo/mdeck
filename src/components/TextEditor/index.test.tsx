import React, { useContext } from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { act, renderHook } from "@testing-library/react-hooks";
import userEvent from "@testing-library/user-event";

import {
  MARKDOWN_CONTEXT_DEFAULT_INITIAL_STATE,
  MarkdownContext,
  MarkdownContextProvider,
} from "contexts/MarkdownContext2";
import { TEXT_AREA_LINE_HEIGHT, TextEditor } from ".";
import { useSyncTextArea } from "./useSyncTextArea";
import { useTrackTextAreaScroll } from "./useTrackTextAreaScroll";

describe("<TextEditor />", () => {
  test("should render and match the snapshot", () => {
    const { asFragment } = render(<TextEditor />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("should set md on textarea change", () => {
    render(
      <MarkdownContextProvider>
        <TextEditor />
      </MarkdownContextProvider>
    );
    const textarea = screen.getByRole("textbox");

    expect(textarea.textContent).toBe("");
    userEvent.type(textarea, "# Hello World!");
    expect(textarea).toHaveValue("# Hello World!");
  });

  describe("useSyncTextArea", () => {
    test("should sync textarea to slideshowLineNumber corresponding text", () => {
      const textarea = document.createElement("textarea");
      const ref = { current: textarea };
      const slideshowLineNumber = 7;

      renderHook(() => {
        useSyncTextArea({
          ref,
          slideshowLineNumber,
          textAreaLineHeight: TEXT_AREA_LINE_HEIGHT,
        });
      });

      expect(textarea.scrollTop).toBe(
        TEXT_AREA_LINE_HEIGHT * slideshowLineNumber
      );
    });
  });

  describe("useTrackTextAreaScroll", () => {
    let textarea;
    let lineCount;
    let scrollHeight;

    beforeEach(() => {
      // Fill textarea with lines of text to scroll
      textarea = document.createElement("textarea");
      lineCount = 7;
      scrollHeight = lineCount * TEXT_AREA_LINE_HEIGHT;
      const textValue = Array(...Array(lineCount))
        .map(() => "test")
        .join("\n");
      textarea.value = textValue;
    });

    const wrapper = ({ children }) => (
      <MarkdownContextProvider>{children}</MarkdownContextProvider>
    );

    test("should set textLineNumber to the top most visible line number on scroll if isActive", () => {
      const ref = { current: textarea };
      const { result } = renderHook(
        () => {
          const { dispatch, state } = useContext(MarkdownContext);
          useTrackTextAreaScroll({
            dispatch,
            isActive: true,
            ref,
            textAreaLineHeight: TEXT_AREA_LINE_HEIGHT,
          });
          return state;
        },
        { wrapper }
      );

      expect(result.current.textLineNumber).toBe(
        MARKDOWN_CONTEXT_DEFAULT_INITIAL_STATE.textLineNumber
      );

      act(() => {
        fireEvent.scroll(textarea, { target: { scrollTop: scrollHeight } });
      });

      expect(result.current.textLineNumber).toBe(lineCount);
    });

    test("should not set textLineNumber on scroll if isActive is false", () => {
      const ref = { current: textarea };
      const { result } = renderHook(
        () => {
          const { dispatch, state } = useContext(MarkdownContext);
          useTrackTextAreaScroll({
            dispatch,
            isActive: false,
            ref,
            textAreaLineHeight: TEXT_AREA_LINE_HEIGHT,
          });
          return state;
        },
        { wrapper }
      );

      act(() => {
        fireEvent.scroll(textarea, { target: { scrollTop: scrollHeight } });
      });

      expect(result.current.textLineNumber).toBe(
        MARKDOWN_CONTEXT_DEFAULT_INITIAL_STATE.textLineNumber
      );
    });
  });
});
