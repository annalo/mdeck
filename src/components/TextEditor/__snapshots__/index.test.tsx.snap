// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextEditor /> should render and match the snapshot 1`] = `
<DocumentFragment>
  .c0 {
  background-color: #fafafa;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  background-color: transparent;
  border: none;
  font-size: 15px;
  height: 100%;
  line-height: 18px;
  outline: none;
  padding: 1em;
  resize: none;
  tab-size: 2;
  -ms-overflow-style: none;
  -webkit-scrollbar-width: none;
  -moz-scrollbar-width: none;
  -ms-scrollbar-width: none;
  scrollbar-width: none;
}

.c1::-webkit-scrollbar {
  display: none;
}

<div
    class="c0"
  >
    <textarea
      class="c1"
    />
  </div>
</DocumentFragment>
`;
